scope_shared/scope/documents/document_set.py

```diff
@@ -0,0 +1,399 @@
+# Allow typing to forward reference
+# TODO: Not necessary with Python 3.11
+from __future__ import annotations
+
+import bson.objectid
+from datetime import datetime
+import pytz
+import secrets
+from typing import Dict, Iterable, List, NewType, Optional, Tuple, Union
+
+# This will allow either until we actually have a Document type
+Document = Union[NewType("Document", dict), dict]
+
+DocumentType = NewType('DocumentType', str)
+SetId = NewType('SetId', str)
+DocumentKey = Union[Tuple[DocumentType], Tuple[DocumentType, SetId]]
+
+
+def datetime_from_document(
+    *,
+    document: dict,
+) -> datetime:
+    return bson.objectid.ObjectId(document["_id"]).generation_time.astimezone(pytz.utc)
+
+
+def document_id_from_datetime(
+    generation_time: datetime,
+) -> str:
+    hex_datetime = str(bson.objectid.ObjectId.from_datetime(
+        generation_time=generation_time
+    ))[0:8]
+
+    hex_random = secrets.token_hex(8)
+
+    return hex_datetime + hex_random
+
+
+def document_key(
+    *,
+    document: Document
+) -> DocumentKey:
+    """
+    Obtain a key for the provided document.
+    """
+
+    key = (
+        document["_type"],
+    )
+    if "_set_id" in document:
+        key = (
+            document["_type"],
+            document["_set_id"],
+        )
+
+    return key
+
+
+class DocumentSet:
```

Comment:
Thoughts on reordering class members for better readability? 

Maybe in the following order: magic methods, properties, private methods, and public methods.

Reply:
They're currently sort of naively alphabetical.  Is there a style guide on this?